src/test/java/jenkins/plugins/git/AbstractGitSCMSourceTest.java

```diff
@@ -426,6 +427,11 @@ public void retrieveRevision() throws Exception {
         assertNull(fileAt("\n", run, source, listener));
         assertThat(source.fetchRevisions(listener), hasItems("master", "dev", "v1"));
         // we do not care to return commit hashes or other references
+        // Commit hashes which are not heads:
```

Comment:
# 1. This is a false test.

At a minimum we need to blow away the local repo cache in-between testing `fileAt` otherwise we will miss the fetch failure from the revision, e.g. `be86b317fb6c2c8572f14f49422ba92627f835b4` being treated as a branch ref spec, i.e. `refs/heads/be86b317fb6c2c8572f14f49422ba92627f835b4` with its resulting log message of 

```
git fetch origin +refs/heads/be86b317fb6c2c8572f14f49422ba92627f835b4:refs/origin/be86b317fb6c2c8572f14f49422ba92627f835b4
fatal: Couldn't find remote ref refs/heads/be86b317fb6c2c8572f14f49422ba92627f835b4
```

Because this test has already checked `fileAt("dev",...)` when `dev` was `v3` the local cache will already contain e.g. `be86b317fb6c2c8572f14f49422ba92627f835b4`

# 2. This is an incomplete test

We do not cover the case where the `v3` revision is in the remote repository but is not advertised on any head. For that case we would need to

```
sampleRepo.git("reset", "--hard", "HEAD^");
sampleRepo.write("file", "v4");
sampleRepo.git("commit", "--all", "--message=v4"); // dev
```

And, most critically, this is the test case that you have actually almost fixed (it's not fixed because we will also fail on fetch:

```
git fetch origin +refs/heads/be86b317fb6c2c8572f14f49422ba92627f835b4:refs/origin/be86b317fb6c2c8572f14f49422ba92627f835b4
fatal: Couldn't find remote ref refs/heads/be86b317fb6c2c8572f14f49422ba92627f835b4
```

But the change you have made is for the case where the revision is in the remote repository but not advertised on a branch.